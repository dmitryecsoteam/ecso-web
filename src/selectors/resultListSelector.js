// Sorts array by first property. If values are equal - sorts by second
const sort = (arr, first, second, descFirst = true, descSecond = true) => {

    // Short helper function. Returns:
    // 1:  when a > b
    // -1: when a < b
    // 0:  when a = b
    const cmp = (a, b) => (a > b) - (a < b);

    return arr.sort((a, b) => {

        // if first values are equal push null in second values to the end
        if (!cmp(a[first],b[first])) {
            if (!a[second]) return 1;
            if (!b[second]) return -1;
        }

        return (descFirst ? -cmp(a[first],b[first]) : cmp(a[first],b[first])) || (descSecond ? -cmp(a[second],b[second]) : cmp(a[second],b[second]));
    });
}


export default (travels, { sortBy, desc }) => {

    if (sortBy === 'price') {

        // filter non-null and null priceAirplane travels
        const filtered = travels.filter(i => i.priceAirplane);
        const nulls = travels.filter(i => !i.priceAirplane);

        // sort both arrays. sorting by avgRating must be descending (travels with greater rating first)
        sort(filtered, 'priceAirplane', 'avgRating', desc, true);
        sort(nulls, 'priceAirplane', 'avgRating', true, true);

        // append null priceAirplane to non-null
        return filtered.concat(nulls);

    } else if (sortBy === 'relevance') {

        // sorting by priceAirplane must be ascending (low prices first)
        return sort(travels.slice(), 'avgRating', 'priceAirplane', desc, false);
        
    } else {
        return travels;
    }
}