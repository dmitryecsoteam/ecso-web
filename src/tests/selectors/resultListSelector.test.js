import travels from '../fixtures/travels';
import resultListSelector from '../../selectors/resultListSelector';

test('should sort by price desc', () => {
    const result = resultListSelector(travels, { sortBy: 'price', desc: true });
    expect(result).toMatchSnapshot();
});

test('should sort by price asc', () => {
    const result = resultListSelector(travels, { sortBy: 'price', desc: false });
    expect(result).toMatchSnapshot();
});

test('should sort by relevance desc', () => {
    const result = resultListSelector(travels, { sortBy: 'relevance', desc: true });
    expect(result).toMatchSnapshot();
});

test('should sort by relevance asc', () => {
    const result = resultListSelector(travels, { sortBy: 'relevance', desc: false });
    expect(result).toMatchSnapshot();
});