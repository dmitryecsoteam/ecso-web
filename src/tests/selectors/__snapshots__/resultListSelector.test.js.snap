// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should sort by price asc 1`] = `
Array [
  Object {
    "avgRating": 4.5,
    "priceAirplane": 100,
  },
  Object {
    "_id": "5c6c7e42e3b7e1edb974a39q",
    "avgRating": 4,
    "carDistance": 150,
    "date": "2019-04-22",
    "destination": Object {
      "_id": 3,
      "beachRating": 1,
      "countryEn": "Japan",
      "foodRating": 4,
      "museumRating": 5,
      "nameEn": "Nara",
      "natureRating": 3,
      "nightlifeRating": 2,
      "shoppingRating": 5,
    },
    "priceAirplane": 100,
  },
  Object {
    "avgRating": 4,
    "priceAirplane": 115,
  },
  Object {
    "_id": "5c6c7e42e3b7e1edb974a39f",
    "avgRating": 3.5,
    "carDistance": 512,
    "date": "2019-04-22",
    "destination": Object {
      "_id": 2,
      "beachRating": 1,
      "countryEn": "Japan",
      "foodRating": 5,
      "museumRating": 4,
      "nameEn": "Osaka",
      "natureRating": 4,
      "nightlifeRating": 5,
      "shoppingRating": 4,
    },
    "priceAirplane": 115,
  },
  Object {
    "avgRating": 5,
    "priceAirplane": 150,
  },
  Object {
    "avgRating": 4,
    "priceAirplane": null,
  },
  Object {
    "avgRating": 1,
    "priceAirplane": null,
  },
]
`;

exports[`should sort by price desc 1`] = `
Array [
  Object {
    "avgRating": 5,
    "priceAirplane": 150,
  },
  Object {
    "avgRating": 4,
    "priceAirplane": 115,
  },
  Object {
    "_id": "5c6c7e42e3b7e1edb974a39f",
    "avgRating": 3.5,
    "carDistance": 512,
    "date": "2019-04-22",
    "destination": Object {
      "_id": 2,
      "beachRating": 1,
      "countryEn": "Japan",
      "foodRating": 5,
      "museumRating": 4,
      "nameEn": "Osaka",
      "natureRating": 4,
      "nightlifeRating": 5,
      "shoppingRating": 4,
    },
    "priceAirplane": 115,
  },
  Object {
    "avgRating": 4.5,
    "priceAirplane": 100,
  },
  Object {
    "_id": "5c6c7e42e3b7e1edb974a39q",
    "avgRating": 4,
    "carDistance": 150,
    "date": "2019-04-22",
    "destination": Object {
      "_id": 3,
      "beachRating": 1,
      "countryEn": "Japan",
      "foodRating": 4,
      "museumRating": 5,
      "nameEn": "Nara",
      "natureRating": 3,
      "nightlifeRating": 2,
      "shoppingRating": 5,
    },
    "priceAirplane": 100,
  },
  Object {
    "avgRating": 4,
    "priceAirplane": null,
  },
  Object {
    "avgRating": 1,
    "priceAirplane": null,
  },
]
`;

exports[`should sort by relevance asc 1`] = `
Array [
  Object {
    "avgRating": 1,
    "priceAirplane": null,
  },
  Object {
    "_id": "5c6c7e42e3b7e1edb974a39f",
    "avgRating": 3.5,
    "carDistance": 512,
    "date": "2019-04-22",
    "destination": Object {
      "_id": 2,
      "beachRating": 1,
      "countryEn": "Japan",
      "foodRating": 5,
      "museumRating": 4,
      "nameEn": "Osaka",
      "natureRating": 4,
      "nightlifeRating": 5,
      "shoppingRating": 4,
    },
    "priceAirplane": 115,
  },
  Object {
    "_id": "5c6c7e42e3b7e1edb974a39q",
    "avgRating": 4,
    "carDistance": 150,
    "date": "2019-04-22",
    "destination": Object {
      "_id": 3,
      "beachRating": 1,
      "countryEn": "Japan",
      "foodRating": 4,
      "museumRating": 5,
      "nameEn": "Nara",
      "natureRating": 3,
      "nightlifeRating": 2,
      "shoppingRating": 5,
    },
    "priceAirplane": 100,
  },
  Object {
    "avgRating": 4,
    "priceAirplane": 115,
  },
  Object {
    "avgRating": 4,
    "priceAirplane": null,
  },
  Object {
    "avgRating": 4.5,
    "priceAirplane": 100,
  },
  Object {
    "avgRating": 5,
    "priceAirplane": 150,
  },
]
`;

exports[`should sort by relevance desc 1`] = `
Array [
  Object {
    "avgRating": 5,
    "priceAirplane": 150,
  },
  Object {
    "avgRating": 4.5,
    "priceAirplane": 100,
  },
  Object {
    "_id": "5c6c7e42e3b7e1edb974a39q",
    "avgRating": 4,
    "carDistance": 150,
    "date": "2019-04-22",
    "destination": Object {
      "_id": 3,
      "beachRating": 1,
      "countryEn": "Japan",
      "foodRating": 4,
      "museumRating": 5,
      "nameEn": "Nara",
      "natureRating": 3,
      "nightlifeRating": 2,
      "shoppingRating": 5,
    },
    "priceAirplane": 100,
  },
  Object {
    "avgRating": 4,
    "priceAirplane": 115,
  },
  Object {
    "avgRating": 4,
    "priceAirplane": null,
  },
  Object {
    "_id": "5c6c7e42e3b7e1edb974a39f",
    "avgRating": 3.5,
    "carDistance": 512,
    "date": "2019-04-22",
    "destination": Object {
      "_id": 2,
      "beachRating": 1,
      "countryEn": "Japan",
      "foodRating": 5,
      "museumRating": 4,
      "nameEn": "Osaka",
      "natureRating": 4,
      "nightlifeRating": 5,
      "shoppingRating": 4,
    },
    "priceAirplane": 115,
  },
  Object {
    "avgRating": 1,
    "priceAirplane": null,
  },
]
`;
